# Explore here

importacion de las librerias 

In [ ]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns  
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso
from sklearn.metrics import r2_score, mean_squared_error
import joblib  

Se descarga la data del url 


In [ ]:


url = "https://breathecode.herokuapp.com/asset/internal-link?id=733&path=demographic_health_data.csv"

df = pd.read_csv(url)
df

In [ ]:
df.shape

In [ ]:
df.info()

In [ ]:
df.select_dtypes(include="object").columns

In [ ]:
df = df.drop(columns=['COUNTY_NAME', 'STATE_NAME'])

In [ ]:
df.select_dtypes(include="object").columns

In [ ]:
df.isnull().sum()

In [ ]:
corr = df.corr()["diabetes_prevalence"].sort_values(ascending=False)
corr

In [ ]:
strong_corr = corr[abs(corr) > 0.6]
strong_corr

In [ ]:
column_drop = [

    "diabetes_Upper 95% CI",
    "diabetes_Lower 95% CI", 
    "CKD_Lower 95% CI",
    "CKD_Upper 95% CI",  
    "anycondition_Lower 95% CI",
    "Heart disease_Lower 95% CI", 
    "anycondition_Upper 95% CI",  
    "COPD_Lower 95% CI",
    "Heart disease_Upper 95% CI",
    "COPD_Upper 95% CI",
    "Obesity_Lower 95% CI",
    "Obesity_Upper 95% CI",
    "MEDHHINC_2018",
]

df =df.drop(columns=column_drop)


In [ ]:
corr = df.corr(numeric_only=True)
corr["diabetes_prevalence"].sort_values(ascending=False)

In [ ]:


plt.hist(df["CKD_prevalence"], bins=30)
plt.xlabel("CKD_prevalence")
plt.ylabel("Frecuencia")
plt.title("Distribucion de CKD_prevalence")
plt.show()

In [ ]:
X = df[[
    "CKD_prevalence",
    "Heart disease_prevalence",
    "COPD_prevalence",
    "Obesity_prevalence",
    "anycondition_prevalence",
    "PCTPOVALL_2018",
    "Median_Household_Income_2018",
    "Percent of adults with less than a high school diploma 2014-18",
    "Percent of adults with a bachelor's degree or higher 2014-18"
]]

y = df["diabetes_prevalence"]

In [ ]:


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [ ]:


regre = LinearRegression()
regre.fit(X_train, y_train)
y_pred =regre.predict(X_test)
r2 = r2_score(y_test,y_pred)
mse = mean_squared_error(y_test, y_pred)

print("r2: ",r2)
print("MSE: ", mse)

In [ ]:
coef_df = X_train.columns.to_frame(name="feature")
coef_df["coef"] = regre.coef_
coef_df.sort_values(by="coef", ascending=False)

In [ ]:


lasso = Lasso(alpha=0.1)  
lasso.fit(X_train, y_train)
y_pred_lasso = lasso.predict(X_test)

r2_lasso = r2_score(y_test, y_pred_lasso)
print("R2 Lasso:", r2_lasso)

In [ ]:


alphas = np.linspace(0.01, 20, 50) 
r2_scores = []  


In [ ]:


for a in alphas:
    lasso = Lasso(alpha=a, max_iter=10000) 
    lasso.fit(X_train, y_train)
    y_pred = lasso.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    r2_scores.append(r2)

In [ ]:


plt.figure(figsize=(10,6))
plt.plot(alphas, r2_scores, marker='o', color='blue')
plt.xlabel("Alpha")
plt.ylabel("R2 en test")
plt.title("R2 vs Alpha en Lasso")
plt.grid(True)
plt.show()

In [ ]:
best_idx = np.argmax(r2_scores)
best_alpha = alphas[best_idx]

lasso_best = Lasso(alpha=best_alpha, max_iter=10000)
lasso_best.fit(X_train, y_train)

coef_df = X_train.columns.to_frame(name="feature")
coef_df["coef"] = lasso_best.coef_
coef_df.sort_values(by="coef", ascending=False)
coef_df

In [ ]:
best_idx = np.argmax(r2_scores)
best_alpha = alphas[best_idx]

print("Mejor alpha:", best_alpha)
print("R² correspondiente:", r2_scores[best_idx])

In [ ]:


lasso_best = Lasso(alpha=best_alpha, max_iter=10000)
lasso_best.fit(X_train, y_train)

In [ ]:


coef_df = pd.DataFrame({
    "feature": X_train.columns,
    "coef": lasso_best.coef_
}).sort_values(by="coef", ascending=False)

coef_df

In [ ]:
y_pred_lasso = lasso_best.predict(X_test)

In [ ]:
r2 = r2_score(y_test, y_pred_lasso)
mse = mean_squared_error(y_test, y_pred_lasso)
rmse = np.sqrt(mse)

print("R² en test:", r2)
print("MSE en test:", mse)
print("RMSE en test:", rmse)

In [30]:
y_pred_lr = regre.predict(X_test)

r2_lr = r2_score(y_test, y_pred_lr)
rmse_lr = np.sqrt(mean_squared_error(y_test, y_pred_lr))

print("R² LinearRegression:", r2_lr)
print("RMSE LinearRegression:", rmse_lr)

R² LinearRegression: 0.9470021644422953
RMSE LinearRegression: 0.6528932872938026


In [ ]:


joblib.dump(lasso_best, "../models/lasso_diabetes_model.pkl")



Después de entrenar el modelo, me di cuenta de que la regresión lineal funciona bastante bien para este problema. Obtuve un R2 de 0.94, lo que significa que el modelo logra explicar casi el 95% de la variación en la prevalencia de diabetes, lo cual es un resultado bastante alto.

Además, el RMSE fue de 0.65, lo que indica que el error promedio de predicción es bajo, así que el modelo no solo explica bien los datos, sino que también predice con buena precisión.

En general, considero que el modelo es sólido y demuestra que las variables relacionadas con enfermedades crónicas y factores socioeconómicos tienen una relación fuerte con la diabetes.